### PR TITLE
remove the 'dereferenceable' attribute from Box

### LINKED ISSUE #66600: `Box` is marked as "dereferenceable" for the duration of the call
This function
```rust
pub fn foo(x: Box<i32>) { drop(x); }
```
compiles to
```
define void @_ZN10playground3foo17h15d47dec4ef032baE(i32* noalias align 4 dereferenceable(4)) unnamed_addr #1 !dbg !148 {
start:
  %x = alloca i32*, align 8
  store i32* %0, i32** %x, align 8
  call void @llvm.dbg.declare(metadata i32** %x, metadata !150, metadata !DIExpression()), !dbg !151
  %1 = load i32*, i32** %x, align 8, !dbg !152, !nonnull !4
; call core::mem::drop
  call void @_ZN4core3mem4drop17had227526e86e8e2bE(i32* noalias align 4 dereferenceable(4) %1), !dbg !153
  br label %bb1, !dbg !153

bb1:                                              ; preds = %start
  ret void, !dbg !154
}
```
Notice the `dereferenceable` attribute! Under current LLVM semantics, this means "dereferenceable for the entire duration of this function body". That is, clearly, not accurate.

This issue is closely related to https://github.com/rust-lang/rust/issues/55005, but affects *all* `Box` instead of just a few uses of references, so I felt it is a separate discussion.

I propose we remove the `dereferencable` attribute from `Box` for now. It seems like the situation might [improve with future LLVM versions](https://github.com/rust-lang/rust/issues/55005#issuecomment-545633772), but we should first make things sound.

Thanks to @HadrienG2 for pointing this out. Cc @rust-lang/wg-unsafe-code-guidelines

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/66600

r? @eddyb @rkruppe

### CODE REVIEW

**Comment by RalfJung** [comment]
@rust-timer queue

**Comment by RalfJung** [comment]
@Mark-Simulacrum this said "finished benchmarking" but there are no results at the comparison URL yet?

**Comment by bjorn3** [comment]
@RalfJung It does now. It takes some time before the parent commit gets benchmarked after `@rust-timer` says it has finished benchmarking the child.

**Comment by RalfJung** [comment]
I see.

Timings look like noise. But this is not necessarily a good benchmark.

File: src/librustc_target/abi/call/mod.rs [hunk 1]
```diff
@@ -69,6 +69,7 @@ mod attr_impl {
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 pub struct ArgAttributes {
     pub regular: ArgAttribute,
+    /// The dereferenceable size of the pointee.
```
Comment (eddyb) [inline src/librustc_target/abi/call/mod.rs]: `dereferenceable` is LLVM jargon, if you want to document this I'd write something more like:
> The minimum size of the pointee, guaranteed to be valid for the duration of the whole call (corresponding to LLVM's `dereferenceable` and `dereferenceable_or_null` attributes).

File: src/librustc/ty/layout.rs [hunk 2]
```diff
@@ -2526,9 +2526,15 @@ where
 
             if let Some(pointee) = layout.pointee_info_at(cx, offset) {
                 if let Some(kind) = pointee.safe {
-                    attrs.pointee_size = pointee.size;
                     attrs.pointee_align = Some(pointee.align);
 
+                    // `Box` (`UniqueBorrowed`) are not necessarily dereferencable
+                    // for the entire duration of the function, so set their size to 0.
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: This should probably mention the deallocation aspect that the test does.

File: src/librustc/ty/layout.rs [hunk 3]
```diff
@@ -2526,9 +2526,15 @@ where
...
+                    attrs.pointee_size =  match kind {
```
Comment (aclonegeek) [inline src/librustc/ty/layout.rs]:
```suggestion
                    attrs.pointee_size = match kind {
```

File: src/librustc_target/abi/call/mod.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_target/abi/call/mod.rs]: I'd use the same jargon in an eventual Rust spec, but sure.

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
r=me if there are no concerns left (e.g. perf)

**Comment by RalfJung** [comment]
@pnkfelix so how do we proceed? I guess we should nominate this for T-lang and T-compiler discussion?

**Comment by pnkfelix** [comment]
(I marked as rollup=never because its not totally certain that there won't be performance regressions associated with this, and I'd rather pay the cost of an extra bors cycle up front than have to track that down in a rollup or have it otherwise masked by unrelated improvements in a rollup.)

**Pull request merged**